includes/utils/class-amp-cache-pool.php

```diff
@@ -0,0 +1,170 @@
+	 */
+	public function get( $key ) {
+		return $this->is_object_cache_available
+			? wp_cache_get( $key, $this->group )
+			: $this->get_rotated_transient( "{$this->group}-{$key}" );
+	}
+
+	/**
+	 * Store a value under a given key in the cache.
+	 *
+	 * @param string $key   Key under which to store the value.
+	 * @param mixed  $value Value to store in the cache.
+	 */
+	public function set( $key, $value ) {
+		if ( $this->is_object_cache_available ) {
+			wp_cache_set( $key, $value, $this->group );
+		} else {
+			$this->set_rotated_transient( "{$this->group}-{$key}", $value );
+		}
+	}
+
+	/**
+	 * Get a value from a rotating transient pool.
+	 *
+	 * @param string $key Key of the value to get.
+	 * @return mixed Value for the requested key.
+	 */
+	private function get_rotated_transient( $key ) {
+		$pool_index = array_search( $key, self::$pool_maps[ $this->group ], true );
+
+		if ( false === $pool_index ) {
+			return false;
+		}
+
+		return get_transient( "{$this->group}-pool-slot-{$pool_index}" );
+	}
+
+	/**
+	 * Store a value in the rotating transient pool under a given key.
+	 *
+	 * @param string $key   Key under which to store the value.
+	 * @param mixed  $value Value to store under the given key.
+	 */
+	private function set_rotated_transient( $key, $value ) {
+		$pool_index = array_search( $key, self::$pool_maps[ $this->group ], true );
+
+		// We already have the provided key and the value seems unchanged.
+		if ( false !== $pool_index && $value === self::$pool_maps[ $this->group ][ $pool_index ] ) {
+			return;
+		}
+
+		// As we didn't find the key, we create a new pool slot to store it.
+		if ( false === $pool_index || -1 === self::$pool_indexes[ $this->group ] ) {
+			$this->advance_pool_index();
+			self::$pool_maps[ $this->group ][ self::$pool_indexes[ $this->group ] ] = $key;
+		}
+
+		// The expiration is to ensure transients don't stick around forever
+		// since no LRU flushing like with external object cache.
+		set_transient( "{$this->group}-pool-slot-" . self::$pool_indexes[ $this->group ], $value, MONTH_IN_SECONDS );
+
+		$this->persist_pool_meta();
+	}
+
+	/**
+	 * Read the pool meta information that was persisted.
+	 */
+	private function read_pool_meta() {
+		self::$pool_maps[ $this->group ]    = get_transient( "{$this->group}-pool-map" ) ?: [];
+		self::$pool_indexes[ $this->group ] = get_transient( "{$this->group}-pool-index" ) ?: -1;
+	}
+
+	/**
+	 * Persist the pool meta information.
+	 */
+	private function persist_pool_meta() {
+		set_transient( "{$this->group}-pool-map", self::$pool_maps[ $this->group ] );
+		set_transient( "{$this->group}-pool-index", self::$pool_indexes[ $this->group ] );
+	}
+
+	/**
+	 * Advance the index into the pool cache.
+	 */
+	private function advance_pool_index() {
+		self::$pool_indexes[ $this->group ]++;
+
+		if ( self::$pool_indexes[ $this->group ] >= $this->size ) {
+			self::$pool_indexes[ $this->group ] = 0;
```

Comment:
In regards to https://github.com/ampproject/amp-wp/issues/2092#issuecomment-533004679, this would be where we could potentially show an error message to the user, warning the that stylesheet cache pool is overflowing, yes? This should issue a PHP warning to the error log?

Reply:
No, this is only expressing the "distance travelled". Velocity is distance over time, we don't know anything about the time here. If we hit the end of the pool several times a day, we have a very high velocity. If we only hit it every 3 months, it might just be a site with lots of edits.